lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,549 @@
+    for (auto *op : toDelete) {
+      op->dropAllUses();
+      op->erase();
+    }
+
+    // Reduce max memory consumption and make sure the values cannot be accessed
+    // anymore because we deleted the ops above.
+    state.clear();
+    interned.clear();
+
+    return success();
+  }
+
+private:
+  const ElaborationOptions &options;
+  std::mt19937 rng;
+
+  // A map used to intern elaborator values. We do this such that we can
+  // compare pointers when, e.g., computing set differences, uniquing the
+  // elements in a set, etc. Otherwise, we'd need to do a deep value comparison
+  // in those situations.
+  // Use a pointer as the key with custom MapInfo because of object slicing when
+  // inserting an object of a derived class of ElaboratorValue.
+  // The custom MapInfo makes sure that we do a value comparison instead of
+  // comparing the pointers.
+  DenseMap<ElaboratorValue *, std::unique_ptr<ElaboratorValue>, InternMapInfo>
+      interned;
+
+  // A map from SSA values to a pointer of an interned elaborator value.
+  DenseMap<Value, ElaboratorValue *> state;
+
+  SymbolTable symTable;
+};
+
+//===----------------------------------------------------------------------===//
+// Elaborator Pass
+//===----------------------------------------------------------------------===//
+
+class ElaborationPass : public rtg::impl::ElaborationPassBase<ElaborationPass> {
+public:
+  ElaborationPass() : ElaborationPassBase() {}
+  ElaborationPass(const ElaborationPass &other) : ElaborationPassBase(other) {}
+  ElaborationPass(const rtg::ElaborationOptions &options)
+      : ElaborationPassBase(), options(options) {
+    if (options.seed.has_value())
+      seed = options.seed.value();
+    debugMode = options.debugMode;
+  }
+  void runOnOperation() override;
+  void cloneTargetsIntoTests();
+
+private:
+  ElaborationOptions options;
+  Pass::Option<unsigned> seed{*this, "seed",
+                              llvm::cl::desc("Seed for the RNG.")};
+  Pass::Option<bool> debugMode{*this, "debug",
+                               llvm::cl::desc("Enable debug mode."),
+                               llvm::cl::init(false)};
+};
+} // end namespace
+
+void ElaborationPass::runOnOperation() {
+  if (seed.hasValue())
+    options.seed = seed;
+  options.debugMode = debugMode;
+
+  cloneTargetsIntoTests();
+
+  auto moduleOp = getOperation();
+  SymbolTable table(moduleOp);
+  Elaborator elaborator(table, options);
+
+  for (auto testOp : moduleOp.getOps<TestOp>())
+    if (failed(elaborator.elaborate(testOp)))
+      return signalPassFailure();
+}
+
+void ElaborationPass::cloneTargetsIntoTests() {
+  auto moduleOp = getOperation();
+  for (auto target : llvm::make_early_inc_range(moduleOp.getOps<TargetOp>())) {
+    for (auto test : moduleOp.getOps<TestOp>()) {
+      // If the test requries nothing from a target, we can always run it.
+      if (test.getTarget().getEntryTypes().empty())
+        continue;
```

Comment:
So if a test does not have any requirements we won't run it on each target, just once in general?

Reply:
I don't really know whats best, what you had originally is probably good!